### ISSUE TITLE
Model Agnostic Meta Learning Algorithms

### ISSUE THREAD

**Issue by howardyclo** [howardyclo/papernotes#54]
# Outline

## Model-Agnostic Meta-Learning for Fast Adaptation of Deep Networks (MAML)
- Authors: Chelsea Finn, Pieter Abbeel, Sergey Levine
- Organization: UC Berkeley& OpenAI
- Conference: ICML 2017
- Paper: https://arxiv.org/abs/1703.03400

## On First-Order Meta-Learning Algorithms (FOMAML; Reptile)
- Authors: Alex Nichol and Joshua Achiam and John Schulman
- Organization: OpenAI
- Paper: https://arxiv.org/abs/1803.02999

## How to Train Your MAML
- Authors: Antreas Antoniou, Harrison Edwards, Amos Strokey
- Organization: University of Edinburgh & OpenAI
- Conference: ICLR 2019
- Paper: https://arxiv.org/abs/1810.09502

**Comment by howardyclo** [comment]
## MAML
- A model-agnostic (only gradient decent) meta learning algorithm aims to find a good initialization point for model such that it can be easily fine-tuned (update few steps or even 1 step) on new tasks.
- Shows STOA on few-shot image classification, regression and fast fine-tuning for policy gradient.

### How does it work?
![](https://github.com/howardyclo/papernotes/blob/master/images/MAML/MAML_1.png?raw=true)
- Sample a "task" (with few training and "virtual" test data; all data is still actually from training data)
- Train and test model on the sampled training task.
- Minimizing the virtual test errors of the sampled tasks is our meta-training objective.
- After several meta-training iterations, we meta-test the model using held-out test task just as same as meta-training.
- The loss can be cross-entropy loss for classification, MSE for regression and reward for RL.

**Comment by howardyclo** [comment]
## Reptile
- Introduce a new FOMAML algorithm, Reptile, which works by repeatedly sampling a task, training on it, and **moving the initialization towards the trained weights**  on that task.

## How does it work?
![](https://github.com/howardyclo/papernotes/blob/master/images/MAML/Reptile_1.png?raw=true)
- The U^{k}_{T} means we take k gradient updates in the sampled task T; ϵ = 1/α, i.e. learning rate.
- We can update in batch version (n = number of tasks):
    ![](https://github.com/howardyclo/papernotes/blob/master/images/MAML/Reptile_2.png?raw=true)
- If we only take k = 1 update, this is essentially SGD on the expected loss.
- If we take k > 1 updates, it is not. This will converges differently which considers second-and-higher derivatives.
